src/goto-symex/symex_main.cpp

```diff
@@ -73,14 +73,47 @@ void symex_transition(
     framet &frame = state.call_stack().top();
     const goto_programt::instructiont &instruction=*to;
     for(const auto &i_e : instruction.incoming_edges)
+    {
       if(i_e->is_goto() && i_e->is_backwards_goto() &&
          (!is_backwards_goto ||
           state.source.pc->location_number>i_e->location_number))
       {
         const auto loop_id =
           goto_programt::loop_id(state.source.function_id, *i_e);
         frame.loop_iterations[loop_id].count = 0;
+        frame.loop_iterations[loop_id].is_active = true;
+      }
+    }
+
+    // And conversely, if we're transitioning out of a loop make sure
+    // we state that it's not active anymore.
+
+    // Match the do-while pattern.
```

Comment:
Again note in comments that this is dubious and should be replaced with a proper loop analysis